Doc/howto/concurrency.rst

```diff
@@ -0,0 +1,1451 @@
+threads are involved.
+
+The big tradeoff is that the bytecode interpreter, which executes your
+Python code, only runs while holding the :term:`!GIL`.  That means only
+one thread can be running Python code at a time.  Threads will take
+short turns, so none have to wait too long, but it still prevents
+any actual parallelism.
+
+At the same time, the Python runtime (and extension modules) can
+release the :term:`!GIL` when the thread is going to be doing something
+unrelated to Python, particularly something slow or long,
+like a blocking IO operation.
+
+There is also an ongoing effort to eliminate the :term:`!GIL`:
+:pep:`630`.  Any attempt to remove the :term:`!GIL` necessarily involves
+some slowdown to single-threaded performance and extra maintenance
+burden to the Python project and extension module maintainers.
+However, there is sufficient interest in unlocking full multi-core
+parallelism to justify the current experiment.
+
+.. currentmodule:: None
+
+.. _python-isolated-threads:
+
+Isolated Threads (CSP/Actor Model)
+----------------------------------
+
+There's a major alternative to free-threading, both for multi-core
+parallelism and for a simpler conceptual model: use multiple interpreters.
+
+Python's major implementation, CPython, has for decades supported
+running with multiple independent copies of the Python runtime
+("interpreter") in a single process.  However, these interpreters
+weren't completely isolated from one another; most importantly they
+shared the one :term:`!GIL`.  Over several years a lot of work went
+into improving the isolation between interpreters, culminating in
+no longer sharing a single :term:`!GIL`.
+
+Besides unlocking full multi-core parallelism, the isolation between
+interpreters means that, from a conceptual level, concurrency can be
+simpler.  An interpreter encapsulates all of Python's runtime state,
+including things like :data:`sys.modules`.  By default, interpreters
+mostly don't share any data (including objects) at all.  Anything that
+gets shared is done on a strictly opt-in basis.  That means programmers
+don't need to worry about possible `races <concurrency-downsides_>`_
+with *any* data in the program.  They only need to worry about data
+that was explicitly shared.
+
+Using multiple interpreters is fairly straight-forward:
+
+1. create a new interpreter
+2. switch the current thread to use that interpreter
+3. call :func:`exec`, but targeting the new interpreter
+4. switch back
+
+You can use the :mod:`!interpreters` module (more on that in a moment)
+to do this::
+
+    import interpreters
+
+    script = """if True:
+        # Do something.
+        ...
+        """
+
+    interp = interpreters.create()
+    interp.exec(script)
+
+Note that no threads were involved.  That's because running in an
+interpreter happens relative to the current thread.  New threads
+aren't implicitly involved.  They can be added in explicitly though.
+Why?  For multi-core parallelism.
```

Comment:
```suggestion
Note that no threads were involved because running in an
interpreter happens relative to the current thread.  New threads
aren't implicitly involved, but they can be added explicitly if the
developer wants multi-core parallelism.
```